source/util/huffman_codec.h

```diff
@@ -83,52 +105,98 @@ class HuffmanCodec {
         break;
       }
 
-      Node* left = queue.top();
+      const uint32_t left = queue.top();
       queue.pop();
 
       // Combine left and right into a new tree and push it into the queue.
-      Node* parent = CreateNode();
-      parent->weight = right->weight + left->weight;
-      parent->left = left;
-      parent->right = right;
+      const uint32_t parent = CreateNode();
+      MutableWeightOf(parent) = WeightOf(right) + WeightOf(left);
+      MutableLeftOf(parent) = left;
+      MutableRightOf(parent) = right;
       queue.push(parent);
     }
 
     // Traverse the tree and form encoding table.
     CreateEncodingTable();
   }
 
+  // Creates Huffman codec from saved tree structure.
+  // |nodes| is the list of nodes of the tree, nodes[0] being NIL.
+  // |root_handle| is the index of the root node.
+  HuffmanCodec(uint32_t root_handle, std::vector<Node>&& nodes) {
+    nodes_ = std::move(nodes);
+    assert(!nodes_.empty());
+    assert(root_handle > 0 && root_handle < nodes_.size());
+    assert(!LeftOf(0) && !RightOf(0));
+
+    root_ = root_handle;
+
+    // Traverse the tree and form encoding table.
+    CreateEncodingTable();
+  }
+
+  // Serializes the codec in the following text format:
+  // (<root_handle>, {
+  //   {0, 0, 0},
+  //   {val1, left1, right1},
+  //   {val2, left2, right2},
+  //   ...
+  // })
+  std::string SerializeToText(int indent_num_whitespaces) const {
+    const bool value_is_text = std::is_same<Val, std::string>::value;
+
+    const std::string indent1 = std::string(indent_num_whitespaces, ' ');
+    const std::string indent2 = std::string(indent_num_whitespaces + 2, ' ');
+
+    std::stringstream code;
+    code << "(" << root_ << ", {\n";
+
+    for (const Node& node : nodes_) {
+      code << indent2 << "{";
+      if (value_is_text)
+        code << "\"";
```

Comment:
This text encoding scheme is fine for simple cases.  If you want to deserialize it for subtle cases then you have to use more complicated escaping.  For example, if the node names themselves contain quote characters.
But this is fine for your current uses.